text/0000-router-service.md

```diff
@@ -0,0 +1,440 @@
+
+```js
+Ember.Helper.extend({
+  compute([routeName], { routeInfo }) {
+    return !!routeInfo.find(info => info.name === routeName);
+  }
+}).reopenClass({
+  getDynamicVariables: ['routeInfo']
+});
+```
+
+A more complete version that also matches models and queryParams can be written in the same way. 
+
+2. We can improve `link-to` so that it always finds implicit model arguments from the local context, rather than trying to locate them on the global router service. This will fix longstanding bugs like https://github.com/ember-animation/liquid-fire/issues/347 and it will make it easier to test components that contain `{{link-to}}`. This would also open the door to relative link-tos.
+
+3. `liquid-outlet` can be implemented entirely via public API. It would become:
+
+```hbs
+{{#liquid-bind (get-dynamic-variable "routeInfo") as |currentRouteInfo|}}
+  {{#with-dynamic-variable "routeInfo" currentRouteInfo}}
+    {{outlet}}
+  {{/with-dynamic-variable}}
+{{/liquid-bind}}
+```
+
+4. Prerendering of non-current routes becomes possible. You can use `recognizeAndLoad` to obtain a `RouteInfoWithAttributes` and then use `{{#with-dynamic-variable "routeInfo" myRouteInfo}} {{outlet}} {{/with-dynamic-variable}}` to render it.
+
+
+# Drawbacks
+
+This RFC deprecates only two public extension hooks API, so the API-churn burden may appear low. However, we know that use of the private APIs we're deliberately disabling is widespread, so users will experience churn. We can provide our usual deprecation cycle to give them early warning, but it still imposes some cost.
+
+This RFC doesn't attempt to change the existing and fairly rich semantics for initiating transitions. For example, you can pass either models or IDs, and those have subtle semantic differences. I think an ideal rewrite would also change the semantics of the route hooks and transitionTo to simplify that area.
+
+# Alternatives
+
+## Less Churn
+
+We could adopt some of the existing broadly used APIs as de-facto public. This avoids churn, but imposes a complexity burden on every new learner, who needs to be told "this is a weird API, but it's what we're stuck with".
+
+## Semver Lawyering
+
+I'm interpreting router.js's public/private documentation as out-of-scope for Ember's semver. The fact that we pass an instance of router.js's Transition as our `transition` argument is not documented. An alternative interpretation is that we need to continue supporting those methods marked as public in router.js's docs.
+
+## Optional Helpers
+
+I didn't propose shipping `is-active` and `url-for` template helpers -- I merely showed that they're easy to build using the router service. But we should arguably just ship them as part of the framework too.
+
+## Branching Route Hierarchies
```

Comment:
This is something that I want to revisit; I had a conversation with @machty this summer about his stack routing research. It seems likely we want to try and solve this coincidentally to "rip the bandaid off" so to say.